### PR TITLE
Support middleware for controller actions

### PR DESCRIPTION
**Pull request by alexpeachey**
@alexdavid 

Added support for putting a `middleware` key on your controller class.

If specified, it will add the listed middleware into the route handler chain.

```
middle = (req, res, next) ->
  req.custom = 'foo'
  next()

class UsersController
  middleware:
    index: middle

  index:   (req, res) -> res.end req.custom

module.exports = UsersController
```

Supports single or arrays of middleware for each entry. See feature for example of single and array usage.

### CODE REVIEW

File: src/controller_initializer.coffee [hunk 1]
```diff
@@ -32,5 +32,13 @@ class ControllerInitializer
       controllerName = camelCase Controller.name.replace /Controller$/, ''
       @controllers[controllerName] = new Controller @dependencies
 
+  # Finds any middleware defined for the controller action
+  middlewareFor: ({controllerName, actionName}) ->
+    controller = @controllers[camelCase controllerName]
+    return [] unless controller?
+    return [] unless typeof controller[actionName] is 'function'
+    return [] unless controller.middleware?[actionName]?
+    [].concat controller.middleware[actionName]
```
Comment (alexdavid) [inline src/controller_initializer.coffee:42]: Awesome! So this means we can pass in an array of middleware instead of just a single one if needed?

Comment (alexpeachey) [inline src/controller_initializer.coffee:42]: Yeah, it takes a single or an array. I thought that would be a nicer API. You can see it in action in the feature specs.

File: test/feature/middleware_support.feature [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      module.exports = UsersController
+      """
+    And an exprestive app using defaults
+    When making a <REQUEST> request to "<URL>"
+    Then the response body should be "<RESPONSE BODY>"
+
+    Examples:
+      | REQUEST | URL                  | RESPONSE BODY |
+      | GET     | /users               | foo           |
+
```
Comment (alexdavid) [inline test/feature/middleware_support.feature:30]: Can you add an extra empty new line here? I like having a double line break between scenarios

File: test/feature/middleware_support.feature [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        req.custom1 = 'foo'
+        next()
+
+      middle2 = (req, res, next) ->
+        req.custom2 = 'bar'
+        next()
+
+      class UsersController
+        middleware:
+          index: [middle1, middle2]
```
Comment (alexdavid) [inline test/feature/middleware_support.feature:49]: I love this, this is really cool!

File: src/controller_initializer.coffee [hunk 4]
```diff
@@ -32,5 +32,13 @@ class ControllerInitializer
       controllerName = camelCase Controller.name.replace /Controller$/, ''
       @controllers[controllerName] = new Controller @dependencies
```
Comment (alexdavid) [inline src/controller_initializer.coffee:34]: I've been using 2 lines between methods

**Comment by alexdavid** [comment]
This looks great @alexpeachey! It's a :ship: from me. @charlierudolph could you take a look at this as well and let me know what you think about it?

We should probably also document this new functionality. Either in this PR or I can open a new one and do it

**Comment by alexpeachey** [comment]
I'll fix the spacing, and yes, I can document it. I wanted to make sure people were happy with it before updating the README.

File: test/feature/middleware_support.feature [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        index:   (req, res) -> res.end req.custom
+      module.exports = UsersController
+      """
+    And an exprestive app using defaults
+    When making a <REQUEST> request to "<URL>"
+    Then the response body should be "<RESPONSE BODY>"
+
+    Examples:
+      | REQUEST | URL                  | RESPONSE BODY |
+      | GET     | /users               | foo           |
```
Comment (charlierudolph) [inline test/feature/middleware_support.feature]: The middle column can be shortened. Same for below

File: test/feature/middleware_support.feature [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      """
+      middle = (req, res, next) ->
+        req.custom = 'foo'
+        next()
+
+      class UsersController
+        middleware:
+          index: middle
+
+        index:   (req, res) -> res.end req.custom
```
Comment (charlierudolph) [inline test/feature/middleware_support.feature]: Please remove the spacing between the color and the first paren, same for below

**Comment by charlierudolph** [comment]
Looks good to me

**Comment by alexpeachey** [comment]
Fixed up the spacing issues and updated the README.
@alexdavid and @charlierudolph do you want to make sure the docs look ok before this gets shipped in?

File: test/feature/middleware_support.feature [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        index: (req, res) -> res.end req.custom
+      module.exports = UsersController
+      """
+    And an exprestive app using defaults
+    When making a <REQUEST> request to "<URL>"
+    Then the response body should be "<RESPONSE BODY>"
+
+    Examples:
+      | REQUEST | URL     | RESPONSE BODY |
+      | GET     | /users  | foo           |
```
Comment (charlierudolph) [inline test/feature/middleware_support.feature]: Ultra nitpick - 
There is still one extra space at the end of the middle column. Same as below.

Need to look into using [cucumber_lint](https://github.com/charlierudolph/cucumber_lint) for js projects. It automatically fixes tables formatting for you.

**Comment by charlierudolph** [comment]
Documentation looks good to me.

File: test/feature/middleware_support.feature (hunk 7 above)
Comment (alexdavid) [inline test/feature/middleware_support.feature]: Or you could use vim with [godlygeek/tabular](https://github.com/godlygeek/tabular) and these tables auto-align themselves to fit their content perfectly :)

(Props to @kevgo for showing me this awesome vim plugin)

**Comment by alexdavid** [comment]
:ship: from me

**Comment by alexpeachey** [comment]
Because I originally cloned from my fork and then re-homed origin to here, I wasn't tracking this origin/master so my ship bombed out AFTER shipping, but before the cleanup. I'm closing this and deleting the branch.

**Comment by charlierudolph** [comment]
If you add 

```
closes #<PR number>
```

into the commit summary it closes the PR for you and creates a link to the commit

**Pull request closed without merging**